File: Serenity.Script.UI/Editor/EnumEditor.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Editor, DisplayName("Enumeration"), OptionsType(typeof(EnumEditorOptions))]
    [Element("<input type=\"hidden\"/>")]
    public class EnumEditor : Select2Editor<EnumEditorOptions, Select2Item>, IStringValue
    {
        public EnumEditor(jQueryObject hidden, EnumEditorOptions opt)
            : base(hidden, opt)
        {
            UpdateItems();
        }

        protected virtual void UpdateItems()
        {
            ClearItems();

            var enumType = options.EnumType ?? EnumTypeRegistry.Get(options.EnumKey);

            var enumKey = options.EnumKey;
            if (enumKey == null && enumType != null)
            {
                var enumKeyAttr = enumType.GetCustomAttributes(typeof(EnumKeyAttribute), false);
                if (enumKeyAttr.Length > 0)
                    enumKey = enumKeyAttr[0].As<EnumKeyAttribute>().Value;
            }

            foreach (var x in Enum.GetValues(enumType))
            {
                var name = Enum.ToString(enumType, x.As<Enum>());
                AddItem(((int)x).ToString(), Q.TryGetText("Enums." + enumKey + "." + name) ?? name);
            }
        }
    }

    [Serializable, Reflectable]
    public class EnumEditorOptions
    {
        public EnumEditorOptions()
        {
        }

        [DisplayName("Enum Type Key")]
        public string EnumKey { get; set; }

        [DisplayName("Enum Type Key"), Hidden]
        public Type EnumType { get; set; }
    }
}

File: src/Serenity.Scripts/CoreLib/UI/Editors/EnumEditor.ts
namespace Serenity {
    
    export interface EnumEditorOptions extends Select2CommonOptions {
        enumKey?: string;
        enumType?: any;
    }

    @registerEditor('Serenity.EnumEditor')
    export class EnumEditor extends Select2Editor<EnumEditorOptions, Select2Item> {
        constructor(hidden: JQuery, opt: EnumEditorOptions) {
            super(hidden, opt);
            this.updateItems();
        }

        protected updateItems(): void {
            this.clearItems();

            var enumType = this.options.enumType || EnumTypeRegistry.get(this.options.enumKey);
            var enumKey = this.options.enumKey;

            if (enumKey == null && enumType != null) {
                var enumKeyAttr = getAttributes(enumType, EnumKeyAttribute, false);
                if (enumKeyAttr.length > 0) {
                    enumKey = enumKeyAttr[0].value;
                }
            }

            var values = Enum.getValues(enumType);
            for (var x of values) {
                var name = Enum.toString(enumType, x);
                this.addOption(parseInt(x, 10).toString(),
                    coalesce(tryGetText('Enums.' + enumKey + '.' + name), name), null, false);
            }
        }

        protected allowClear() {
            return (this.options.allowClear ?? true);
        }
    }
}